### PR TITLE
General cleanup for client disconnect handling.

### PR DESCRIPTION
**Pull request by jkotalik**
- Remove extra call to SetHandler in disconnecthandler.
- Lock around setting and getting the pManagedHttpContext to guarantee a disconnect notification is triggered if disconnect happens.
- Call AbortIO on threadpool.

### CODE REVIEW

File: src/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp [hunk 1]
```diff
@@ -108,8 +109,16 @@ IN_PROCESS_HANDLER::NotifyDisconnect()
...
+        m_pDisconnectHandler(m_pManagedHttpContext);
+    }
+    else
+    {
+        // set an atomic bool
+        m_disconnectWithoutCallingManaged = true;
+    }
 
-    assert(m_pManagedHttpContext != nullptr);
     m_pDisconnectHandler(m_pManagedHttpContext);
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp]: delete this.

File: src/Microsoft.AspNetCore.Server.IIS/Core/IISHttpContext.IO.cs [hunk 2]
```diff
@@ -211,7 +212,7 @@ public void Abort(Exception reason)
 
         internal void ConnectionReset()
         {
-            AbortIO();
+            ThreadPool.QueueUserWorkItem(t => AbortIO());
```
Comment (pakrym) [inline src/Microsoft.AspNetCore.Server.IIS/Core/IISHttpContext.IO.cs]: Only dispatch TCS cancellation

File: src/Microsoft.AspNetCore.Server.IIS/Core/IISHttpServer.cs [hunk 3]
```diff
@@ -206,14 +206,18 @@ private static void OnDisconnect(IntPtr pvManagedHttpContext)
             IISHttpContext context = null;
             try
             {
-                context = (IISHttpContext)GCHandle.FromIntPtr(pvManagedHttpContext).Target;
+                var gcHandle = GCHandle.FromIntPtr(pvManagedHttpContext);
+                if (!gcHandle.IsAllocated)
```
Comment (pakrym) [inline src/Microsoft.AspNetCore.Server.IIS/Core/IISHttpServer.cs]: Why?

File: src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISDeployer.cs [hunk 4]
```diff
@@ -364,6 +365,7 @@ private void Stop()
                                 wp.State == WorkerProcessState.Running ||
                                 wp.State == WorkerProcessState.Stopping))
                         {
+                            Logger.LogInformation($"WorkerProcess not stopped yet");
```
Comment (pakrym) [inline src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISDeployer.cs]: Why do we need this in addition to exceptions.

File: test/WebSites/InProcessWebSite/Startup.cs [hunk 5]
```diff
@@ -324,6 +324,7 @@ private async Task ReadAndWriteEcho(HttpContext ctx)
                 result = await ctx.Request.Body.ReadAsync(readBuffer, 0, readBuffer.Length);
             }
         }
+
```
Comment (pakrym) [inline test/WebSites/InProcessWebSite/Startup.cs]: revert

File: src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISDeployer.cs (hunk 4 above)
Comment (jkotalik) [reply src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISDeployer.cs]: Was useful for debugging why the worker process wasn't being killed.

File: src/Microsoft.AspNetCore.Server.IIS/Core/IISHttpServer.cs (hunk 3 above)
Comment (jkotalik) [reply src/Microsoft.AspNetCore.Server.IIS/Core/IISHttpServer.cs]: If the gcHandle is null, we will do a needless throw/catch of a NullReferenceException.

File: src/AspNetCoreModuleV2/AspNetCore/DisconnectHandler.cpp [hunk 6]
```diff
@@ -29,12 +30,23 @@ void DisconnectHandler::NotifyDisconnect()
...
+    IREQUEST_HANDLER* pHandler = nullptr;
+    {
+        SRWExclusiveLock lock(m_handlerLock);
+
+        handler.swap(m_pHandler);
+        pHandler = m_pHandler.get();
+    }
+
+    // SetHandler is called in the destructor of the DisconnectHandler;
+    // don't check the httpContext
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/AspNetCore/DisconnectHandler.cpp]: REmove comment and put pHandler != nullptr first

File: src/AspNetCoreModuleV2/AspNetCore/DisconnectHandler.h [hunk 7]
```diff
@@ -11,15 +11,16 @@ class ASPNET_CORE_PROXY_MODULE;
...
+    DisconnectHandler(IHttpConnection* pHttpConnection)
+        : m_pHandler(nullptr), m_pHttpConnection(pHttpConnection), m_disconnectFired(false)
     {
         InitializeSRWLock(&m_handlerLock);
     }
 
     virtual
     ~DisconnectHandler()
     {
+        //m_pHttpConnection = nullptr;
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/AspNetCore/DisconnectHandler.h]: ?

File: src/AspNetCoreModuleV2/AspNetCore/DisconnectHandler.h [hunk 8]
```diff
@@ -30,10 +31,12 @@ class DisconnectHandler final: public IHttpConnectionStoredContext
     CleanupStoredContext() noexcept override;
 
     void
-    SetHandler(std::unique_ptr<IREQUEST_HANDLER, IREQUEST_HANDLER_DELETER> handler) noexcept;
+    SetHandler(std::unique_ptr<IREQUEST_HANDLER, IREQUEST_HANDLER_DELETER> handler);
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/AspNetCore/DisconnectHandler.h]: has to be noexcept, it's called from the destructor.

File: src/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp [hunk 9]
```diff
@@ -92,16 +94,33 @@ REQUEST_NOTIFICATION_STATUS IN_PROCESS_HANDLER::ServerShutdownMessage()
...
+    m_moduleTracer.RequestDisconnect();
+
     if (m_pApplication->QueryBlockCallbacksIntoManaged() ||
         m_fManagedRequestComplete)
     {
         return;
     }
 
-    m_moduleTracer.RequestDisconnect();
+    // For client disconnects, we are allowed to call m_pDisconnectHandler() multiple times,
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp]: Are we ever going to call m_pDisconnectHandler() multiple times with the current logic we have?

File: src/Microsoft.AspNetCore.Server.IIS/Core/IISHttpServer.cs (hunk 3 above)
Comment (pakrym) [reply src/Microsoft.AspNetCore.Server.IIS/Core/IISHttpServer.cs]: Why do we ever call OnDisconnect with null handle? I though new logic in native should've fixed it.

File: test/Common.FunctionalTests/ClientDisconnectStress.cs [hunk 10]
```diff
@@ -4,6 +4,7 @@
 using System;
 using System.Collections.Generic;
 using System.Net.Http;
+using System.Threading;
```
Comment (pakrym) [inline test/Common.FunctionalTests/ClientDisconnectStress.cs]: revert

**Comment by jkotalik** [comment]
@pakrym we green now. Would like to get this in by tomorrow.

**Review by pakrym — APPROVED** [review]

**Comment by pakrym** [comment]
Good job!

**Pull request merged**